crates/rattler_installs_packages/src/index/package_sources.rs

```diff
@@ -0,0 +1,108 @@
+    /// This URL will be used for all packages by default
+    pub fn new(base_index_url: Url) -> Self {
+        Self {
+            base_source: base_index_url,
+            extra_sources: Default::default(),
+            overrides: Default::default(),
+        }
+    }
+
+    /// Add another index URL
+    pub fn with_index(&mut self, alias: String, url: Url) -> &mut Self {
+        self.extra_sources.push(PackageSource { alias, url });
+        self
+    }
+
+    /// Add an override for a specific package. This will cause the package to be installed
+    /// from the given source and from that source only
+    pub fn with_override(&mut self, package: NormalizedPackageName, alias: String) -> &mut Self {
+        self.overrides.insert(package, alias);
+        self
+    }
+
+    /// Finalize the builder and create a `PackageSources` instance
+    pub fn build(&self) -> miette::Result<PackageSources> {
+        let mut extra_sources_map = BTreeMap::new();
+        self.extra_sources
+            .iter()
+            .enumerate()
+            .map(|(i, source)| {
+                // Offset by one because index_urls will include the default source
+                (source.alias.clone(), i + 1)
+            })
+            .try_for_each(|(alias, index)| {
+                if extra_sources_map.insert(alias.clone(), index).is_some() {
+                    miette::bail!("duplicate index alias '{}'", alias);
+                }
+                Ok(())
+            })?;
+
+        let mut artifact_to_index = BTreeMap::new();
+        self.overrides.iter().try_for_each(|(package, source)| {
+            let index = *extra_sources_map
+                .get(source)
+                .ok_or_else(|| miette::miette!("unknown index alias '{}'", source))?;
+
+            if artifact_to_index.insert(package.clone(), index).is_some() {
+                miette::bail!("duplicate package-source map entry '{}'", package);
+            }
+
+            Ok(())
+        })?;
+
+        let index_urls = std::iter::once(self.base_source.clone())
+            .chain(self.extra_sources.iter().map(|source| source.url.clone()))
+            .collect::<Vec<_>>();
+
+        Ok(PackageSources {
+            index_urls,
+            artifact_to_index,
+        })
+    }
+}
+
+/// A collection of package sources and source overrides.
+/// See [`PackageSourcesBuilder`] for creating an instance of this type.
+pub struct PackageSources {
+    index_urls: Vec<Url>,
+    artifact_to_index: BTreeMap<NormalizedPackageName, usize>,
+}
+
+impl PackageSources {
+    /// Get the index URL for a package
+    pub fn index_url(&self, package: &NormalizedPackageName) -> Vec<&Url> {
+        let maybe_index = self
+            .artifact_to_index
+            .get(package)
+            .map(|&index| &self.index_urls[index]);
+
+        if let Some(url) = maybe_index {
+            vec![url]
+        } else {
+            self.index_urls.iter().collect()
```

Comment:
Should this return all `urls`? That means it will also include any specific package urls right?

Reply:
I guess now that we have a seperate object to handle the indexes for packages we could add both usecases right? Have default indexes as well as package specific ones? 

Not sure if thats useful. Lets leave as is, we can always easily add that later.